qiskit/transpiler/passes/mapping/apply_layout.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+A pass for transforming a circuit with virtual qubits into a circuit with physical qubits.
+"""
+from qiskit.circuit import QuantumRegister
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class ApplyLayout(TransformationPass):
+    """
+    Transforms a DAGCircuit with virtual qubits into a DAGCircuit with physical qubits
+    by applying the Layout given in `property_set`.
+    Requires either of passes to set/select Layout, e.g. `SetLayout`, `TrivialLayout`.
+    """
+
+    def run(self, dag):
+        """
+        Runs the ApplyLayout pass on `dag`.
+        Args:
+            dag (DAGCircuit): DAG to map.
+        Returns:
+            DAGCircuit: A mapped DAG (with physical qubits).
+        Raises:
+            TranspilerError: if no layout is found in `property_set`.
+        """
+        layout = self.property_set["layout"]
+        if not layout:
+            raise TranspilerError(
+                "No 'layout' found in property_set. Please run a Layout pass in advance.")
+
+        n_physical_qubits = 1 + max(layout.get_physical_bits())
+        q = QuantumRegister(n_physical_qubits, 'q')
```

Comment:
@ajavadia I removed `coupling_map`, and tried to switch to `len(layout)` simply, but found some difficulty in the case like `layout = Layout({v[0]: 2, v[1]: 1})` because current DAG cannot subscript int as qubit. Could you check the lines around here and the test case `test_more_physical_qubits_than_virtual_qubits`?

Reply:
I updated the flow in the comment above, and i think it will look like this for a small example. So this pass does not need CouplingMap, because the circuit has already been enlarged to match the full device.

original circuit
v[0] --x----
v[1] --.----

after ancilla allocation + enlargewithancilla:
v[0] --x----
v[1] --.----
ancilla0 -----
ancilla1 -----
ancilla2 -----
ancilla3 -----

after applylayout:
0------ ancilla2
1 ------ ancilla0
2 --x--- v[0]
3 --.--- v[1]
4 ------ ancilla3
5 ------ ancilla1

Reply:
@ajavadia @kdk I changed to assume full layout (i.e. require FullAncillaAllocation), but I couldn't find the way to avoid from enlarging qregs, i.e. doing the same thing as EnlargeWithAncilla within ApplyLayout. Is it OK?

Should I use self.requires to ensure full layout? (If so, coupling will be needed in `__init__`.)
```
        self.requires.append(FullAncillaAllocation(coupling))
```

Reply:
I think what you have done is fine. I'm not sure we need to require the ancilla allocation pass. The error that you raise (`"The 'layout' must be full (with ancilla)."`) probably is enough.